lib/aiken/trees/mt.ak

```diff
@@ -0,0 +1,504 @@
+
+  rh == root_hash
+}
+
+test root_hash_2() {
+  let items = []
+  let mt = from_list(items)
+
+  let node_hash = #""
+
+  root_hash(mt) == node_hash
+}
+
+/// Function atests whether two Merkle Tress are equal, this is the case when their root hashes match.
+pub fn is_equal(self1: MerkleTree, self2: MerkleTree) -> Bool {
+  root_hash(self1) == root_hash(self2)
+}
+
+/// Function returns a total numbers of leaves in the tree.
+pub fn size(self: MerkleTree) -> Int {
+  when self is {
+    Empty -> 0
+    Leaf{..} -> 1
+    Node { left, right, .. } -> size(left) + size(right)
+  }
+}
+
+test size_1() {
+  let items = []
+  let mt = from_list(items)
+  size(mt) == 0
+}
+
+test size_2() {
+  let dog = sha2_256(from_string("dog"))
+  let cat = sha2_256(from_string("cat"))
+  let mouse = sha2_256(from_string("mouse"))
+
+  let mt = from_list([dog, cat, mouse])
+  size(mt) == 3
+}
+
+test size_3() {
+  let dog = sha2_256(from_string("dog"))
+  let cat = sha2_256(from_string("cat"))
+  let mouse = sha2_256(from_string("mouse"))
+  let items = [dog, cat, mouse]
+
+  let mt = from_list(items)
+  size(mt) == 3
+}
+
+test size_4() {
+  let dog = sha2_256(from_string("dog"))
+  let cat = sha2_256(from_string("cat"))
+  let mouse = sha2_256(from_string("mouse"))
+  let horse = sha2_256(from_string("horse"))
+  let pig = sha2_256(from_string("pig"))
+  let bull = sha2_256(from_string("bull"))
+
+  let items = [dog, cat, mouse, horse, pig, bull]
+
+  let mt = from_list(items)
+  size(mt) == 6
+}
+
+fn combine_hash(h1: Hash<alg, a>, h2: Hash<alg, a>) -> Hash<alg, a> {
+  sha2_256(bytearray.concat(h1, h2))
+}
+
+/// Function that returns whether merkle tree has any elements
+pub fn is_empty(self: MerkleTree) -> Bool {
+  when self is {
+    Empty -> True
+    _ -> False
+  }
+}
+
+test is_empty_1() {
+  let mt = Empty
+
+  is_empty(mt)
+}
+
+test is_empty_2() {
+  let dog = sha2_256(from_string("dog"))
+
+  let mt = Leaf { hash: dog }
+
+  is_empty(mt) == False
+}
+
+fn do_proof(
+  self: MerkleTree,
+  item: Hash<Sha2_256, ByteArray>,
+  proof: Proof,
+) -> Option<Proof> {
+  when self is {
+    Empty -> None
+    Leaf { hash } ->
+      if hash == item {
+        Some(proof)
+      } else {
+        None
+      }
+    Node { left, right, .. } -> {
+      let rh = root_hash(right)
+      let lh = root_hash(left)
+      let go_left: Option<Proof> =
+        do_proof(left, item, list.push(proof, Right { value: rh }))
+      when go_left is {
+        Some(_) -> do_proof(left, item, list.push(proof, Right { value: rh }))
+        None -> do_proof(right, item, list.push(proof, Left { value: lh }))
+      }
+    }
+  }
+}
+
+/// Construct a membership 'Proof' from an element and a 'MerkleTree'. Returns
+/// 'None' if the element isn't a member of the tree to begin with.
+/// Note function will return Some([]) in case root of the tree is also it's only one and only element
+pub fn get_proof(
+  self: MerkleTree,
+  item: Hash<Sha2_256, ByteArray>,
```

Comment:
The item ought to be a `ByteArray` here. Or possibly, _anything_ given that we have `serializeData` as a builtin now. I am thinking that we could actually have generalized merkle trees.